File: rofunc/planning/src/robotics-codes-from-scratch-master/matlab/iLQR_manipulator_obstacle.m
%%	  iLQR applied to a planar manipulator task with obstacles avoidance
%% 	  (viapoints task with position+orientation including bounding boxes on f(x))
%%
%%    Copyright (c) 2021 Idiap Research Institute, http://www.idiap.ch/
%%    Written by Sylvain Calinon <https://calinon.ch>
%%
%%    This file is part of RCFS.
%%
%%    RCFS is free software: you can redistribute it and/or modify
%%    it under the terms of the GNU General Public License version 3 as
%%    published by the Free Software Foundation.
%%
%%    RCFS is distributed in the hope that it will be useful,
%%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
%%    GNU General Public License for more details.
%%
%%    You should have received a copy of the GNU General Public License
%%    along with RCFS. If not, see <http://www.gnu.org/licenses/>.

function iLQR_manipulator_obstacle

%% Parameters
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
param.dt = 1E-2; %Time step size
param.nbData = 50; %Number of datapoints
param.nbIter = 50; %Number of iterations for iLQR
param.nbPoints = 1; %Number of viapoints
param.nbObstacles = 2; %Number of obstacles
param.nbVarX = 3; %State space dimension (q1,q2,q3)
param.nbVarU = 3; %Control space dimension (dq1,dq2,dq3)
param.nbVarF = 3; %Objective function dimension (x1,x2,o)
param.l = [2, 2, 1]; %Robot links lengths
param.sz2 = [.5, .8]; %Size of obstacles
param.q = 1E2; %Tracking weight term
param.q2 = 1E0; %Obstacle avoidance weight term
param.r = 1E-3; %Control weight term

param.Mu = [[3; -1; 0]]; %Viapoints (x1,x2,o)

param.Mu2 = [[2.8; 2.0; pi/4], [3.5; .5; -pi/6]]; %Obstacles (x1,x2,o)
% param.Mu2 = [[-1.2; 1.5; pi/4], [-0.5; 2.5; -pi/6]]; %Obstacles (x1,x2,o)
for t=1:param.nbObstacles
	param.A2(:,:,t) = [cos(param.Mu2(3,t)), -sin(param.Mu2(3,t)); sin(param.Mu2(3,t)), cos(param.Mu2(3,t))]; %Orientation
	param.R2(:,:,t) = param.A2(:,:,t) * diag(param.sz2); %Eigendecomposition of covariance matrix S2=R2*R2'
	param.U2(:,:,t) = param.A2(:,:,t) * diag(param.sz2.^-1); %Eigendecomposition of precision matrix Q2=U2*U2'
end

R = speye((param.nbData-1) * param.nbVarU) * param.r; %Control weight matrix (at trajectory level)

%Time occurrence of viapoints
tl = linspace(1, param.nbData, param.nbPoints+1);
tl = round(tl(2:end));
idx = (tl - 1) * param.nbVarX + [1:param.nbVarX]';


%% Iterative LQR (iLQR)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
u = zeros(param.nbVarU*(param.nbData-1), 1); %Initial commands
x0 = [3*pi/4; -pi/2; -pi/4]; %Initial robot pose

%Transfer matrices (for linear system as single integrator)
Su0 = [zeros(param.nbVarX, param.nbVarX*(param.nbData-1)); tril(kron(ones(param.nbData-1), eye(param.nbVarX)*param.dt))];
Sx0 = kron(ones(param.nbData,1), eye(param.nbVarX));
Su = Su0(idx,:);

for n=1:param.nbIter
	x = reshape(Su0 * u + Sx0 * x0, param.nbVarX, param.nbData); %System evolution
	[f, J] = f_reach(x(:,tl), param); %Tracking objective
	
	[f2, J2, id2] = f_avoid(x, param); %Avoidance objective
	Su2 = Su0(id2,:);
	
	du = (Su' * J' * J * Su * param.q + Su2' * J2' * J2 * Su2 * param.q2 + R) \ ...
			(-Su' * J' * f(:) * param.q - Su2' * J2' * f2(:) * param.q2 - u * param.r); %Gradient
		
	%Estimate step size with backtracking line search method
	alpha = 1;
	cost0 = norm(f(:))^2 * param.q + norm(f2(:))^2 * param.q2 + norm(u)^2 * param.r;
	while 1
		utmp = u + du * alpha;
		xtmp = reshape(Su0 * utmp + Sx0 * x0, param.nbVarX, param.nbData);
		ftmp = f_reach(xtmp(:,tl), param); %Tracking objective
		ftmp2 = f_avoid(xtmp, param); %Avoidance objective
		cost = norm(ftmp(:))^2 * param.q + norm(ftmp2(:))^2 * param.q2 + norm(utmp)^2 * param.r;
		if cost < cost0 || alpha < 1E-3
			break;
		end
		alpha = alpha * 0.5;
	end
	u = u + du * alpha;
	
	if norm(du * alpha) < 1E-2
		break; %Stop iLQR when solution is reached
	end
end
disp(['iLQR converged in ' num2str(n) ' iterations.']);

%Log data
r.x = x;
r.f = [];
for j=param.nbVarU:-1:1
	r.f = [r.f; fkin(x(1:j,:), param.l(1:j))];
end
r.u = reshape(u, param.nbVarU, param.nbData-1);


%% Plot state space
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
figure('position',[10,10,800,800],'color',[1,1,1]); hold on; axis off;
ftmp = fkin0(x(:,1), param.l);
plot(ftmp(1,:), ftmp(2,:), '-','linewidth',4,'color',[.8 .8 .8]);
ftmp = fkin0(x(:,tl(end)), param.l);
plot(ftmp(1,:), ftmp(2,:), '-','linewidth',4,'color',[.4 .4 .4]);
al = linspace(-pi, pi, 50);
for t=1:param.nbObstacles
	msh = param.R2(:,:,t) * [cos(al); sin(al)] + repmat(param.Mu2(1:2,t), 1, 50);
	patch(msh(1,:), msh(2,:), [.8 .8 .8],'linewidth',2,'edgecolor',[.4 .4 .4]);
end
for j=param.nbVarU:-1:1
	plot(r.f((j-1)*param.nbVarF+1,:), r.f((j-1)*param.nbVarF+2,:), '-','linewidth',1,'color',[.2 .2 .2]);
	plot(r.f((j-1)*param.nbVarF+1,:), r.f((j-1)*param.nbVarF+2,:), '.','markersize',10,'color',[.2 .2 .2]);
end
plot(r.f(1,:), r.f(2,:), '-','linewidth',2,'color',[0 0 0]);
plot(r.f(1,:), r.f(2,:), '.','markersize',10,'color',[0 0 0]);
plot(r.f(1,[1,tl]), r.f(2,[1,tl]), '.','markersize',30,'color',[0 0 0]);
axis equal; 

pause(10);
end 


%%%%%%%%%%%%%%%%%%%%%%
%Logarithmic map for R^2 x S^1 manifold
function e = logmap(f, f0)
	e(1:2,:) = f(1:2,:) - f0(1:2,:);
	e(3,:) = imag(log(exp(f0(3,:)*1i)' .* exp(f(3,:)*1i).'));
end

%%%%%%%%%%%%%%%%%%%%%%
%Forward kinematics for end-effector (in robot coordinate system)
function f = fkin(x, L)
	T = tril(ones(length(L)));
	f = [L * cos(T * x); ...
		L * sin(T * x); ...
		mod(sum(x,1)+pi, 2*pi) - pi]; %x1,x2,o (orientation as single Euler angle for planar robot)
end

%%%%%%%%%%%%%%%%%%%%%%
%Forward kinematics for all robot articulations (in robot coordinate system)
function f = fkin0(x, L)
	T = tril(ones(size(x,1)));
	T2 = T * diag(L);
	f = [T2 * cos(T * x), ...
		T2 * sin(T * x)]'; 
	f = [zeros(2,1), f];
end

%%%%%%%%%%%%%%%%%%%%%%
%Jacobian with analytical computation (for single time step)
function J = jkin(x, L)
	T = tril(ones(length(L)));
	J = [-sin(T * x)' * diag(L) * T; ...
		cos(T * x)' * diag(L) * T; ...
		ones(1, length(L))]; %x1,x2,o
end

%%%%%%%%%%%%%%%%%%%%%%
%Cost and gradient for a viapoints reaching task (in object coordinate system)
function [f, J] = f_reach(x, param)
% 	f = fkin(x, param) - param.Mu; %Error by ignoring manifold
	f = logmap(fkin(x, param.l), param.Mu); %Error by considering manifold
	J = []; 
	for t=1:size(x,2)		
		Jtmp = jkin(x(:,t), param.l);
		J = blkdiag(J, Jtmp);
	end
end

%%%%%%%%%%%%%%%%%%%%%%
%Cost and gradient for an obstacle avoidance task
function [f, J, id] = f_avoid(x, param)
	f=[]; J=[]; id=[];
	for i=1:param.nbObstacles
		for j=1:param.nbVarU
			for t=1:param.nbData
				xee = fkin(x(1:j,t), param.l(1:j));
				e = param.U2(:,:,i)' * (xee(1:2) - param.Mu2(1:2,i));
				ftmp = 1 - e' * e; %quadratic form
				%Bounding boxes 
				if ftmp > 0
					f = [f; ftmp];
					Jrob = [jkin(x(1:j,t), param.l(1:j)), zeros(param.nbVarF, param.nbVarU-j)];
					Jtmp = -e' * param.U2(:,:,i)' * Jrob(1:2,:); %quadratic form
					J = blkdiag(J, Jtmp);
					id = [id, (t-1) * param.nbVarU + [1:param.nbVarU]'];
				end
			end
		end
	end
end

File: rofunc/planning/src/robotics-codes-from-scratch-master/python/iLQR_manipulator_obstacle.py
'''
    Batch iLQR with obstacles avoidance applied on a manipulator example

    Copyright (c) 2021 Idiap Research Institute, http://www.idiap.ch/
    Written by Jeremy Maceiras <jeremy.maceiras@idiap.ch>,
    Sylvain Calinon <https://calinon.ch>

    This file is part of RCFS.

    RCFS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 3 as
    published by the Free Software Foundation.

    RCFS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with RCFS. If not, see <http://www.gnu.org/licenses/>.
'''

import numpy as np
import numpy.matlib
import matplotlib.pyplot as plt
import matplotlib.patches as patches

# Helper functions
# ===============================

# Logarithmic map for R^2 x S^1 manifold
def logmap(f,f0):
    position_error = f[:,:2] - f0[:,:2]
    orientation_error = np.imag(np.log( np.exp( f0[:,-1]*1j ).conj().T * np.exp(f[:,-1]*1j).T )).conj().reshape((-1,1))
    error = np.hstack(( position_error , orientation_error ))
    return error

# Forward kinematics for E-E
def fkin(x,ls):
	x = x.T
	A = np.tril(np.ones([len(ls),len(ls)]))
	f = np.vstack((ls @ np.cos(A @ x), 
				   ls @ np.sin(A @ x), 
				   np.mod(np.sum(x,0)+np.pi, 2*np.pi) - np.pi)) #x1,x2,o (orientation as single Euler angle for planar robot)
	return f.T

# Forward Kinematics for all joints
def fkin0(x):
    T = np.tril(np.ones([param["nbVarX"],param["nbVarX"]]))
    T2 = np.tril(np.matlib.repmat(param["linkLengths"],len(x),1))
    f = np.vstack(( 
        T2 @ np.cos(T@x),
        T2 @ np.sin(T@x)
        )).T
    f = np.vstack(( 
        np.zeros(2),
        f
        ))
    return f

def f_avoid(x):
    global param
    f=[]
    idx=[]
    J=np.zeros((0,0))
    
    for i in range(param["nbObstacles"]):
        for j in range(param["nbVarU"]):
            for t in range(param["nbData"]):
                
                f_x = fkin(x[t,:(j+1)],param["linkLengths"][:(j+1)]).flatten()
                e = param["U_obs"][i].T @ (f_x[:2] - param["Obs"][i][:2])
                ftmp = 1 - e.T @ e

                if ftmp > 0:
                    f += [ftmp]
                    J_rbt = np.hstack(( jkin(x[t,:(j+1)] , param["linkLengths"][:(j+1)]), np.zeros(( param["nbVarF"] , param["nbVarU"]-(j+1))) ))
                    J_tmp = (-e.T @ param["U_obs"][i].T @ J_rbt[:2]).reshape((-1,1))
                    #J_tmp = 1*(J_rbt[:2].T @ param["U_obs"][i] @ e).reshape((-1,1))

                    J2 = np.zeros(( J.shape[0] + J_tmp.shape[0] , J.shape[1] + J_tmp.shape[1] ))
                    J2[:J.shape[0],:J.shape[1]] = J
                    J2[-J_tmp.shape[0]:,-J_tmp.shape[1]:] = J_tmp
                    J = J2 # Numpy does not provid a blockdiag function...

                    idx.append( (t-1)*param["nbVarU"] + np.array(range(param["nbVarU"])) )

    f = np.array(f)
    idx = np.array(idx)
    return f,J.T,idx

def jkin(x,ls):
    global param
    T = np.tril( np.ones((len(ls),len(ls))) )
    J = np.vstack((
        -np.sin(T@x).T @ np.diag(ls) @ T,
        np.cos(T@x).T @ np.diag(ls) @ T,
        np.ones(len(ls))
    ))
    return J

# Cost and gradient
def f_reach( x ):
    global param

    f = logmap(fkin(x,param["linkLengths"]),param["mu"])
    J = np.zeros(( len(x) * param["nbVarF"] , len(x) * param["nbVarX"] ))

    for t in range(x.shape[0]):
        Jtmp = jkin(x[t], param["linkLengths"])
        J[ t*param["nbVarF"]:(t+1)*param["nbVarF"] , t*param["nbVarX"]:(t+1)*param["nbVarX"]] = Jtmp
    return f,J

# General param parameters
# ===============================
param = {
"dt":1e-2,
"nbData":50,
"nbIter":20,
"nbPoints":1,
"nbObstacles":2,
"nbVarX":3,
"nbVarU":3,
"nbVarF":3,
"linkLengths":[2,2,1],
"sizeObstacle":[.5,.8],
"Q_track":1e2,
"Q_avoid":1e0,
"r":1e-3
}

# Task parameters
# ===============================

# Targets
param["mu"] = np.asarray([
[3,-1,0] ])				# x , y , orientation


# Obstacles
param["Obs"] = np.array([
[2.8, 2, np.pi/4],        # [x1,x2,o]
[3.5, .5, -np.pi/6]        # [x1,x2,o]
])

param["A_obs"] = np.zeros(( param["nbObstacles"] , 2 , 2 ))
param["S_obs"] = np.zeros(( param["nbObstacles"] , 2 , 2 ))
param["R_obs"] = np.zeros(( param["nbObstacles"] , 2 , 2 ))
param["Q_obs"] = np.zeros(( param["nbObstacles"] , 2 , 2 ))
param["U_obs"] = np.zeros(( param["nbObstacles"] , 2 , 2 )) # Q_obs[t] = U_obs[t].T @ U_obs[t]
for i in range(param["nbObstacles"]):
    orn_t = param["Obs"][i][-1]
    param["A_obs"][i] = np.array([              # Orientation in matrix form
    [ np.cos(orn_t) , -np.sin(orn_t)  ],
    [ np.sin(orn_t) , np.cos(orn_t)]
    ])

    param["S_obs"][i] = param["A_obs"][i] @ np.diag(param["sizeObstacle"])**2 @ param["A_obs"][i].T # Covariance matrix
    param["R_obs"][i] = param["A_obs"][i] @ np.diag(param["sizeObstacle"]) # Square root of covariance matrix

    param["Q_obs"][i] = np.linalg.inv( param["S_obs"][i] ) # Precision matrix
    param["U_obs"][i] = param["A_obs"][i] @ np.diag(1/np.array(param["sizeObstacle"])) # "Square root" of param["Q_obs"][i]


# Regularization matrix
R = np.identity( (param["nbData"]-1) * param["nbVarU"] ) * param["r"]

# System parameters
# ===============================

# Time occurence of viapoints
tl = np.linspace(0,param["nbData"],param["nbPoints"]+1)
tl = np.rint(tl[1:]).astype(np.int64)-1
idx = np.array([ i + np.arange(0,param["nbVarX"],1) for i in (tl* param["nbVarX"])]) 

u = np.zeros( param["nbVarU"] * (param["nbData"]-1) ) # Initial control command
x0 = np.array( [3*np.pi/4 , -np.pi/2 , - np.pi/4] ) # Initial state (in joint space)

# Transfer matrices (for linear system as single integrator)
Su0 = np.vstack([np.zeros((param["nbVarX"], param["nbVarX"]*(param["nbData"]-1))), 
np.tril(np.kron(np.ones((param["nbData"]-1, param["nbData"]-1)), np.eye(param["nbVarX"])*param["dt"]))]) 
Sx0 = np.kron( np.ones(param["nbData"]) , np.identity(param["nbVarX"]) ).T
Su = Su0[idx.flatten()] # We remove the lines that are out of interest

# Solving iLQR
# ===============================

for i in range( param["nbIter"] ):
    x = Su0 @ u + Sx0 @ x0
    x = x.reshape( (  param["nbData"] , param["nbVarX"]) )

    f, J = f_reach(x[tl])  # Tracking objective
    f2, J2, id2 = f_avoid(x)# Avoidance objective

    if len(id2) > 0: # Numpy does not allow zero sized array as Indices
        Su2 = Su0[id2.flatten()]
        du = np.linalg.inv(Su.T @ J.T @ J @ Su * param["Q_track"] + Su2.T @ J2.T @ J2 @ Su2 * param["Q_avoid"] + R) @ \
            (-Su.T @ J.T @ f.flatten() * param["Q_track"] - Su2.T @ J2.T @ f2.flatten() * param["Q_avoid"] - u * param["r"])
    else: # It means that we have a collision free path
        du = np.linalg.inv(Su.T @ J.T @ J @ Su * param["Q_track"] + R) @ \
            (-Su.T @ J.T @ f.flatten() * param["Q_track"] - u * param["r"])
    # Perform line search
    alpha = 1
    cost0 = np.linalg.norm(f.flatten())**2 * param["Q_track"] + np.linalg.norm(f2.flatten())**2 * param["Q_avoid"] + np.linalg.norm(u) * param["r"]

    while True:
        utmp = u + du * alpha
        xtmp = Su0 @ utmp + Sx0 @ x0
        xtmp = xtmp.reshape( (  param["nbData"] , param["nbVarX"]) )
        ftmp, _ = f_reach(xtmp[tl])
        f2tmp,_,_ = f_avoid(xtmp)
        cost = np.linalg.norm(ftmp.flatten())**2 * param["Q_track"] + np.linalg.norm(f2tmp.flatten())**2 * param["Q_avoid"] + np.linalg.norm(utmp) * param["r"]

        if cost < cost0 or alpha < 1e-3:
            u = utmp
            print("Iteration {}, cost: {}, alpha: {}".format(i,cost,alpha))
            break

        alpha /=2
    
    if np.linalg.norm(alpha * du) < 1e-2: # Early stop condition
       break

# Ploting
# ===============================

plt.figure()
plt.axis("off")
plt.gca().set_aspect('equal', adjustable='box')

# Get points of interest
f = fkin(x,param["linkLengths"])

for i in range(param["nbVarX"]-1):
    fi = fkin( x[:,:i+1] , param["linkLengths"][:i+1] )
    plt.plot(fi[:,0],fi[:,1],c="black",alpha=.5,linestyle='dashed')

f00 = fkin0(x[0])
fT0 = fkin0(x[-1])

plt.plot( f00[:,0] , f00[:,1],c='black',linewidth=5,alpha=.2)
plt.plot( fT0[:,0] , fT0[:,1],c='black',linewidth=5,alpha=.6)

plt.plot(f[:,0],f[:,1],c="black",marker="o",markevery=[0]+tl.tolist()) #,label="Trajectory"


# Plot bounding box or via-points
ax = plt.gca()
color_map = ["deepskyblue","darkorange"]
for i in range(param["nbPoints"]):
        plt.scatter(param["mu"][i,0],param["mu"][i,1],s=100,marker="X",c=color_map[i])

# Plot obstactles
al = np.linspace(-np.pi,np.pi,50)
ax = plt.gca()
for i in range(param["nbObstacles"]):
    D,V = np.linalg.eig(param["S_obs"][i])
    D = np.diag(D)
    R = np.real(V@np.sqrt(D+0j))
    msh = (R @ np.array([np.cos(al),np.sin(al)])).T + param["Obs"][i][:2]
    p=patches.Polygon(msh,closed=True)
    ax.add_patch(p)

plt.show()
